### PR TITLE
Stop consuming input in fuzzers to select target environment

### LINKED ISSUE #4450: Adapt SPIR-V fuzzers to take target environment from existing bytes in the binary
A number of the libFuzzer fuzzer targets under test/fuzzers interpret the first word of the given binary as a target environment, like this:

```
  const spv_context context =
      spvContextCreate(*reinterpret_cast<const spv_target_env*>(data));
  if (context == nullptr) return 0;

  data += sizeof(spv_target_env);
  size -= sizeof(spv_target_env);
```

This prevents such fuzzers being equipped with a corpus of SPIR-V files to work on: any corpus would have to be adapted so that some target environment word is prepended to the file.

An alternative would be to have the fuzzers take regular SPIR-V binaries as input, and decide on a target environment based on certain bytes of the file.

For example, assuming there are N target environments that we care about:

```
spv_target_env env;
switch (data[size/2] % N) {
  case 0:
    env = SPV_1_0; // or whatever the right enum is
    break;
  case 1:
    env = SPV_1_1; // or whatever the right enum is
   ...
}
const spv_contexgt context = spvContextCreate(env);
```

This would then allow the fuzzers to run with a corpus.

A detail is that probably most of the bytes in the file will be 0, so it might be better to sum a range of bytes and then mod the result of the sum.

@zoddicus would this change seem reasonable to you?

### PR DESCRIPTION
**Pull request by zoddicus**
Instead calculate a hash based on the input and use that as a seed
into random data generation for the target env.

Also fixes issue where input data was not actually being fed into
one fuzzer.

Fixes #4450

### CODE REVIEW

**Comment by zoddicus** [comment]
Pretty sure the failing tests are not from my patch, since the only non-fuzzer code I added was a utility function

**Comment by alan-baker** [comment]
If you rebase, the unrelated failures should be gone.

**Comment by zoddicus** [comment]
> If you rebase, the unrelated failures should be gone.

Thanks, done

**Comment by zoddicus** [comment]
CI-windows-VS2015-release looks like timeout flake to me, re-running that build

**Comment by zoddicus** [comment]
Could I get a review on this?

File: source/spirv_target_env.cpp [hunk 1]
```diff
@@ -286,6 +286,38 @@ bool spvIsOpenGLEnv(spv_target_env env) {
...
+    case SPV_ENV_OPENGL_4_0:
+    case SPV_ENV_OPENGL_4_1:
+    case SPV_ENV_OPENGL_4_2:
+    case SPV_ENV_OPENGL_4_3:
+    case SPV_ENV_OPENGL_4_5:
+      return true;
+    case SPV_ENV_WEBGPU_0:
+      break;
+  }
+  return false;
```
Comment (afd) [inline source/spirv_target_env.cpp:318]: If this switch covers all the enum cases then I would favour "return false" in the SPV_ENV_WEBGPU_0 branch, and "assert(false && "Unreachable")" at the end.

File: source/spirv_target_env.h [hunk 2]
```diff
@@ -28,6 +28,9 @@ bool spvIsOpenCLEnv(spv_target_env env);
 // Returns true if |env| is an OPENGL environment, false otherwise.
 bool spvIsOpenGLEnv(spv_target_env env);
 
+// Returns true if |env| is an implemented/valid environment, false otherwise.
```
Comment (afd) [inline source/spirv_target_env.h:31]: It's not that clear to me what it means for an environment to be "implemented/valid". I would assume that every value of the enum's range is valid in some sense (otherwise why would it be in the enum range?). Could you elaborate slightly in the comment?

File: test/fuzzers/random_generator.h [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+
+namespace spvtools {
+namespace fuzzers {
+
+/// Pseudo random generator utility class for fuzzing
+class RandomGenerator {
+ public:
+  /// @brief Initializes the internal engine
+  /// @param seed - seed value passed to engine
+  explicit RandomGenerator(uint64_t seed);
```
Comment (afd) [inline test/fuzzers/random_generator.h:30]: If we want to be strict about "include what you use" then to use uint64_t (and similar) you should include <cstdint>. But we're not strict about this in the rest of SPIR-V Tools.

File: test/fuzzers/random_generator.cpp [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+
+/// Helper for obtaining a seed bias value for HashCombine with a bit-width
+/// dependent on the size of size_t.
+template <int SIZE_OF_SIZE_T>
+struct HashCombineOffset {};
+/// Specialization of HashCombineOffset for size_t == 4.
+template <>
+struct HashCombineOffset<4> {
+  /// @returns the seed bias value for HashCombine()
+  static constexpr inline uint32_t value() { return 0x7f4a7c16; }
```
Comment (afd) [inline test/fuzzers/random_generator.cpp:44]: Some indication of why this specific literal is used would be nice, in a comment.

File: test/fuzzers/random_generator.cpp [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+};
+/// Specialization of HashCombineOffset for size_t == 8.
+template <>
+struct HashCombineOffset<8> {
+  /// @returns the seed bias value for HashCombine()
+  static constexpr inline uint64_t value() { return 0x9e3779b97f4a7c16; }
```
Comment (afd) [inline test/fuzzers/random_generator.cpp:50]: Similar.

File: test/fuzzers/random_generator.cpp [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+};
+
+/// HashCombine "hashes" together an existing hash and hashable values.
+template <typename T>
+void HashCombine(size_t* hash, const T& value) {
+  constexpr size_t offset = HashCombineOffset<sizeof(size_t)>::value();
+  *hash ^= std::hash<T>()(value) + offset + (*hash << 6) + (*hash >> 2);
+}
+
+/// Calculate the hash for the contents of a c-style data buffer
```
Comment (afd) [inline test/fuzzers/random_generator.cpp:60]: c -> C

File: test/fuzzers/random_generator.cpp [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+/// @param data - pointer to buffer to be hashed
+/// @param size - number of elements in buffer
+/// @returns hash of the data in the buffer
+size_t HashBuffer(const uint8_t* data, const size_t size) {
+  size_t hash = 102931;
```
Comment (afd) [inline test/fuzzers/random_generator.cpp:65]: Why this specific literal?

Nit: why hex above and decimal here?

File: test/fuzzers/random_generator.cpp [hunk 8]
```diff
@@ -0,0 +1,126 @@
...
+    }
+  }
+
+  spv_target_env target_env = SPV_ENV_WEBGPU_0;
+  while (target_env == SPV_ENV_WEBGPU_0) {
+    target_env =
+        static_cast<spv_target_env>(RandomUInt(&engine_, 0u, range_end));
+  }
+  return target_env;
+}
```
Comment (afd) [inline test/fuzzers/random_generator.cpp:100]: A potential problem with this is that it's an undefined behaviour to cast an integer to an enum if the integer falls out of the enum value's range (if no backing type is specified) - so you'll get a UB when you go past the last element of the range.

Perhaps the backing type for the enum should be changed to uint32_t. Or a special enum value indicating the maximum in the range should be added.

File: test/fuzzers/spvtools_as_fuzzer.cpp [hunk 9]
```diff
@@ -18,18 +18,26 @@
...
 
 extern "C" int LLVMFuzzerTestOneInput(const uint8_t* data, size_t size) {
-  if (size < sizeof(spv_target_env) + 1) return 0;
-
-  const spv_context context =
-      spvContextCreate(*reinterpret_cast<const spv_target_env*>(data));
-  if (context == nullptr) return 0;
+  if (size < 1) {
+    return 0;
+  }
```
Comment (afd) [inline test/fuzzers/spvtools_as_fuzzer.cpp:26]: Should the 0 bytes case really be excluded? A target environment could be chosen totally at random in that case. Similar for the other fuzzer targets.

**Review by afd — CHANGES_REQUESTED** [review]
Looking good! Various nits.

File: test/fuzzers/spvtools_as_fuzzer.cpp (hunk 9 above)
Comment (zoddicus) [reply test/fuzzers/spvtools_as_fuzzer.cpp:26]: Picking an arbitrary random number would eliminate reproducibility of the test case, since a different target env could be picked each time the fuzzer is run with the same input.

The 0 sized case feels like an edge case that is better for something like unit or other functional testing, since it is feels pretty tractable to exhaustively test all of different classes of cases, i.e. {nullptr, 0}, {valid pointer, 0}, {invalid pointer, 0}.

Comment (afd) [reply test/fuzzers/spvtools_as_fuzzer.cpp:26]: I see what you mean about randomization.

I guess I'd favour including the 0-sized case, and using a specific target env in that scenario. I agree its an edge case, but it's the sort of edge case that people forget about with other testing methods, so wouldn't hurt to cover it here.

But I'm fine either way.

File: test/fuzzers/random_generator.cpp [hunk 10]
```diff
@@ -0,0 +1,133 @@
...
+      SPV_ENV_UNIVERSAL_1_3,       SPV_ENV_VULKAN_1_1,
+      SPV_ENV_OPENCL_1_2,          SPV_ENV_OPENCL_EMBEDDED_1_2,
+      SPV_ENV_OPENCL_2_0,          SPV_ENV_OPENCL_EMBEDDED_2_0,
+      SPV_ENV_OPENCL_EMBEDDED_2_1, SPV_ENV_OPENCL_EMBEDDED_2_2,
+      SPV_ENV_OPENCL_2_1,          SPV_ENV_OPENCL_2_2,
+      SPV_ENV_UNIVERSAL_1_4,       SPV_ENV_VULKAN_1_1_SPIRV_1_4,
+      SPV_ENV_UNIVERSAL_1_5,       SPV_ENV_VULKAN_1_2,
+      SPV_ENV_OPENGL_4_0,          SPV_ENV_OPENGL_4_1,
+      SPV_ENV_OPENGL_4_2,          SPV_ENV_OPENGL_4_3,
+      SPV_ENV_OPENGL_4_5};
```
Comment (afd) [inline test/fuzzers/random_generator.cpp:104]: My problem with this is that it requires maintenance: as more versions are added, which they will be over time, someone needs to remember to update this.

If we could add a SPV_ENV_MAX value to the enum this problem would go away.

I see your point about testing code not necessarily being sufficient motivation for this. But actually OSS-Fuzz has found an undefined beahvior in spirv-val related to this problem, where an out of range value is cast to this enum type. With no easy way to know how large the enum type is, it's hard to fix the UB.

So: I'm fine with this for now but I think we should discuss it - I have opened https://github.com/KhronosGroup/SPIRV-Tools/issues/4558.

**Review by afd — APPROVED** [review]
Looks good to me - a couple of follow up comments to consider at your discretion / to discuss after this PR.

File: test/fuzzers/spvtools_as_fuzzer.cpp (hunk 9 above)
Comment (zoddicus) [reply test/fuzzers/spvtools_as_fuzzer.cpp:26]: Changing this to picking universal 1.0 in the 0 case.
